### ISSUE TITLE
[MSHADE-316] Minijar: <excludeDefaults>true</excludeDefaults>

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-shade-plugin#317]
**[Markus Karg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkarg)** opened **[MSHADE-316](https://issues.apache.org/jira/browse/MSHADE-316?redirect=false)** and commented

Minijar currently respects includes, but these have a drawback: Once you defined an include, the _default includes_ are gone! But what you actually want to have when combining minijar with includes is just an "exception" to minijar's filtering, not a complete replacement of the default includes!

So what we need to make this work intuitively is either "explicit minijar exceptions", or a "useDefaultIncludes=true" option that turns the notion from "replace default includes by the single given include" to "add the single given include ontop of the default includes".

---

**Remote Links:**
- [GitHub Pull Request #18
](https://github.com/apache/maven-shade-plugin/pull/18)
- [GitHub Pull Request #19
](https://github.com/apache/maven-shade-plugin/pull/19)

**Comment by jira-importer** [comment]
**[Markus Karg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkarg)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) No, this is a misunderstanding. The idea is **not** to have full control over the classes being added. The idea is (see also my PR) is to **prevent minjar's removal of explicitly marked classes**. Use case: A dependency shall get included into a shaded jar, but that dependency is loaded by **Reflection**, so the original artifact does not directly reference the **Class** to load, but only holds a **String** with the class name. The Minijar filter will remove that class, so the shaded JAR will be broken. With my PR, you can add an _explicit minijar exemption_ on that single class name (or pattern), so minijar will keep that file even **without a technically detectable** need. This cannot be done with the existing include filters, because once _at least one_ include is explictly defined, the default (include _all_) is switched off so _all_ other classes would be _excluded_. The solution is my PR which simply provides an **exemption** for Minijar. There is a rather popular real-life scenario: Minijar breaks _Eclipse Jersey_ because one of Jersey's dependency is loaded by Reflection.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

>  As of version 1.6, minimizeJar will respect classes that were specifically marked for inclusion in a filter. Note that specifying an include filter for classes in an artifact implicitly excludes all non-specified classes in that artifact.

That's the last line on the page and that explains the current behavior. It also kind of indicates what the preferred solution should be:

```xml
<minimizeJar>true</minimizeJar>
<filters>
  <filter>
     <artifact>log4j:log4j</artifact>
     <excludeDefaults>false</excludeDefaults>
     <includes>
         <include>somefile</include>
     </includes>
  </filter>
</filters>
```

[Reporting](https://maven.apache.org/ref/3.6.1/maven-model/maven.html#class_reporting) has a similar option, so I'd prefer to follow that.

**Comment by jira-importer** [comment]
**[Markus Karg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkarg)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) Ok I will create another PR implementing that alternative solution. :-) Just too bad that you didn't tell me earlier as a reaction upon my comment of April 4 where I let you choose your preference and proposed exactly what you now want in your recent comment... ;-(

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Well, the more people that ask for my opinion, the less time I have to really dive into every issue. And it is all spare time, maintaining all Maven projects is voluntary job.

**Comment by jira-importer** [comment]
**[Markus Karg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkarg)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) Fully agree with all you say. But see, it is the same for me, too. So I think you understand that it is not funny to squander one day of volunteer time with the wrong solution.

**Comment by jira-importer** [comment]
**[Markus Karg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkarg)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) Provided PR that implements excludeDefaults=true as you requested. Please review https://github.com/apache/maven-shade-plugin/pull/19. :)

**Comment by jira-importer** [comment]
**[Markus Karg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkarg)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) All requested changes are implemented. Please continue review https://github.com/apache/maven-shade-plugin/pull/19. :)

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [2f6663c725d5b7fd42a177d530eb3577f370f323](https://gitbox.apache.org/repos/asf?p=maven-shade-plugin.git;a=commit;h=2f6663c725d5b7fd42a177d530eb3577f370f323)
Thanks for the patch!
